apisix/upstream.lua

```diff
@@ -52,41 +60,112 @@ end
+                           node.host, ":", port or node.port, " err: ", err)
+        end
+    end
+
+    if upstream.parent then
+        core.table.insert(upstream.parent.clean_handlers, function ()
+            core.log.info("try to release checker: ", tostring(checker))
+            checker:stop()
+        end)
+
+    else
+        core.table.insert(healthcheck_parent.clean_handlers, function ()
+            core.log.info("try to release checker: ", tostring(checker))
+            checker:stop()
+        end)
+    end
+
+    core.log.info("create new checker: ", tostring(checker))
+    return checker
+end
+
+
+local function fetch_healthchecker(upstream, healthcheck_parent, version)
+    if not upstream.checks then
+        return
+    end
+
+    local checker = lrucache_checker(upstream, version,
+                                     create_checker, upstream,
+                                     healthcheck_parent)
+    return checker
+end
+
+
 function _M.set_by_route(route, api_ctx)
     if api_ctx.upstream_conf then
         core.log.warn("upstream node has been specified, ",
                       "cannot be set repeatedly")
-        return true
+        return
     end
 
     local up_id = route.value.upstream_id
     if up_id then
         if not upstreams then
-            return false, "need to create a etcd instance for fetching "
+            return 500, "need to create a etcd instance for fetching "
                           .. "upstream information"
         end
 
         local up_obj = upstreams:get(tostring(up_id))
         if not up_obj then
-            return false, "failed to find upstream by id: " .. up_id
+            return 500, "failed to find upstream by id: " .. up_id
         end
         core.log.info("upstream: ", core.json.delay_encode(up_obj))
 
         local up_conf = up_obj.dns_value or up_obj.value
         set_directly(api_ctx, up_conf.type .. "#upstream_" .. up_id,
                      up_obj.modifiedIndex, up_conf, up_obj)
-        return true
+        return
     end
 
     local up_conf = (route.dns_value and route.dns_value.upstream)
                     or route.value.upstream
     if not up_conf then
-        return false, "missing upstream configuration in Route or Service"
+        return 500, "missing upstream configuration in Route or Service"
+    end
+
+    if up_conf.service_name then
+        if not discovery then
+            return 500, "discovery is uninitialized"
+        end
+        up_conf.nodes = discovery.nodes(up_conf.service_name)
     end
 
     set_directly(api_ctx, up_conf.type .. "#route_" .. route.value.id,
                  api_ctx.conf_version, up_conf, route)
-    return true
+
+    local nodes_count = up_conf.nodes and #up_conf.nodes or 0
+    if nodes_count == 0 then
+        return 502, "no valid upstream node"
+    end
+
+    if nodes_count > 1 then
+        local checker = fetch_healthchecker(up_conf, route,
```

Comment:
I think it might be better to move the creation of health-checker object to the set_directly function，
Sometimes, I call the set_directly function directly in the plugin，proxy to another uptream instead of the configuration in the route.

Reply:
we can add this in new PR ^_^